### ISSUE TITLE
[java] New rule: SuspiciousEqualsCall

### ISSUE THREAD

**Issue by UncleOwen** [pmd/pmd#6715]
**Proposed Rule Name:** SuspiciousEqualsCall? EqualsOnIncompatibleTypes?

**Proposed Category:** Error Prone

**Description:** This rule would trigger on calls of `a.equals(b)`, where the static types of `a` and `b` don't have a common ~supertype (except Object)~ subtype.

Code like this usually comes up when you
* were thinking of slightly different objects (e.g. you meant to call `a.getId().equals(b)` or something like that
* messed up a refactoring.

**Code Sample:** 

```java
Integer i;
String s;
if (i.equals(s)) {    // violation
  // do stuff
}
```

**Possible Properties:**

*  Maybe a stricter mode, where one of the types must be a supertype of the other?

**Comment by zbynek** [comment]
List of possible testcases: https://github.com/pmd/pmd/pull/6006#issuecomment-3217294242

**Comment by zbynek** [comment]
One might also consider lenient mode where checking equality of two types is allowed if they have an intersection.

```java
sealed interface FlatStuff permits Polygon, Circle {}
sealed interface RoundStuff permits Sphere, Circle {}
private record Circle(int radius) {}
private static RoundStuff UNIT_CIRCLE = new Circle(1);

static boolean isUnitCircle(FlatStuff object) {
   return UNIT_CIRCLE.equals(object);
}
```

The use of `equals` above may be against best practices, but is not indication of a bug.

**Comment by UncleOwen** [comment]
The equivalent IntelliJ inspection is [EqualsBetweenInconvertibleTypes](https://www.jetbrains.com/help/inspectopedia/EqualsBetweenInconvertibleTypes.html#locating-this-inspection), where

> One might also consider lenient mode where checking equality of two types is allowed if they have an intersection.

seems to be the default. Maybe we should follow that.

**Comment by UncleOwen** [comment]
I edited the first post, because @zbynek's suggestion makes more sense.

**Comment by zbynek** [comment]
Hm, actually the correct check might look for subtypes OR supertypes -- equality check between a `LinkedList` and `ArrayList` may return true.

**Comment by UncleOwen** [comment]
Hmm, yes, equality check between a `LinkedList` to an `ArrayList` is possible. But not between, say, `Integer` and `Double`. Or between any two classes implementing `Serializable`. Any idea how we can differentiate these cases?

**Comment by UncleOwen** [comment]
I just had a look at what IntelliJ does:

```java
class Foo {
    boolean compareLists(LinkedList<String> ll, ArrayList<String> al) {
        // 0 warnings
        return ll.equals(al);
    }
    
    boolean compareNumbers(Integer i, Double d) {
        // 2 warnings:
        // Result of 'i.equals(d)' is always 'false'
        // 'equals' between objects of inconvertible types 'Integer' and 'Double'
        return i.equals(d);
    }
    
    boolean compareAs(A1 a1, A2 a2) {   // A1 and A2 are custom classes implementing the same interface A (or extend the same class)
        // 1 warning:
        // 'equals' between objects of inconvertible types 'A1' and 'A2' 
        return a1.equals(a2);
    }

    boolean compareIs(I1 i1, I2 i2) {
        // 0 warning if a class exists, that implements both interfaces
        // 1 warning if not.
        // No class found which is a subtype of both 'I1' and 'I2' 
        return i1.equals(i2);
    }
}
```
